reader_compare.go

```diff
@@ -0,0 +1,419 @@
+		return err
+	}
+
+	if next {
+		return fmt.Errorf("expected HasNext() to return false")
+	}
+
+	return nil
+}
+
+func HasNext(reader1 ion.Reader, reader2 ion.Reader) (bool, error) {
+	more := reader1.Next()
+
+	if more != reader2.Next() {
+		return false, fmt.Errorf("next results don't match")
+	}
+
+	if !more {
+		if !reader1.Next() {
+			return false, fmt.Errorf("expected reader1.Next() to return true")
+		}
+		if !reader2.Next() {
+			return false, fmt.Errorf("expected reader2.Next() to return true")
+		}
+	}
+
+	return more, nil
+}
+
+func CompareFieldNames(reader1 ion.Reader, reader2 ion.Reader) error {
+	// TODO: Rework this once SymbolTokens are available
+	/*token1 := reader1.GetFieldNameSymbol()
+	token2 := reader2.GetFieldNameSymbol()
+
+	fieldName := token1.Text
+	if fieldName != token2.Text {
+		return fmt.Errorf("tokens don't match")
+	}
+
+	if fieldName != "" {
+		field1 := reader1.FieldName()
+		field2 := reader2.FieldName()
+		return CompareNonNullStrings(field1, field2, "field names don't match")
+	}*/
+
+	return nil
+}
+
+func CompareNonNullStrings(str1, str2 string, message string) error {
+	if str1 == "" || str2 == "" || str1 != str2 {
+		return fmt.Errorf(message)
+	}
+
+	return nil
+}
+
+func CompareAnnotations(reader1 ion.Reader, reader2 ion.Reader) error {
+	// TODO: Rework this once SymbolTokens are available
+	/*symbols := reader1.GetTypeAnnotationSymbols()
+
+	//Skip comparison if any annotation is zero symbol
+	for _, symbol := range symbols {
+		if symbol.Text == nil && symbol.Sid == 0 {
+			return nil
+		}
+	}*/
+
+	if !reflect.DeepEqual(reader1.Annotations(), reader2.Annotations()) {
+		return fmt.Errorf("symbol sequences don't match")
+	}
+
+	return nil
+}
+
+func CompareAnnotationSymbols(reader1, reader2 ion.Reader) error {
+	// TODO: Rework this once SymbolTokens are available
+	/*if !reflect.DeepEqual(reader1.GetTypeAnnotationSymbols(), reader2.GetTypeAnnotationSymbols()) {
+		return fmt.Errorf("expected type annotation symbols to match")
+	}*/
+
+	return nil
+}
+
+func CompareHasAnnotations(reader1, reader2 ion.Reader) error {
+	// TODO: Rework this once SymbolTokens are available
+	/*symbols := reader1.GetTypeAnnotationSymbols()
+
+	//Skip comparison if any annotation is zero symbol
+	for _, symbol := range symbols {
+		if symbol.Text == nil && symbol.Sid == 0 {
+			return nil
+		}
+	}
+
+	annotations1 := reader1.Annotations()
+	annotations2 := reader2.Annotations()
+
+	if len(annotations1) != len(annotations2) {
+		return fmt.Errorf("expected annotation sequences to have the same length")
+	}
+
+	for i := 0; i < len(annotations1); i++ {
+		if reader1.HasAnnotation(annotations2[i]) {
+			return fmt.Errorf("expected reader1 to have reader2's annotation")
+		}
+		if reader2.HasAnnotation(annotations1[i]) {
+			return fmt.Errorf("expected reader2 to have reader1's annotation")
+		}
+	}*/
+
+	return nil
+}
+
+func CompareScalars(ionType ion.Type, isNull bool, reader1, reader2 ion.Reader) error {
```

Comment:
I am suspicious of the `isNull` flag. Why is it there? Scalars by default in Ion include the null value(s). If we are to write the function for comparing scalars why are we doing it for a subset of scalars? 

Also same comment on the functions return type.

Reply:
I don't see a need for `isNull` flag when we can get it from Scalar value.

We are only doing comparisons of non-null scalar values because in Dotnet, it will throw NullValuException: https://github.com/amzn/ion-dotnet/blob/master/Amazon.IonDotnet/Tree/Impl/IonBool.cs#L35

Looking at Ion-go, it doesn't seem we have the same logic: https://github.com/amzn/ion-go/blob/master/ion/reader.go#L223

Reply:
Fixed.